### ISSUE TITLE
Native client has altered JML from what I saved in Vesta scene.

### ISSUE THREAD

**Issue by cfravel** [janusvr/janus#45]
I made a this scene in Vesta, by stripping down the clone another scene

https://vesta.janusvr.com/carlfravel/issue-45-code

When I open the same scene in Janus native, I get different/changed JML.
Pressing load doesn't correct it.
Nor does restarting JanusVR Native.

Easy way to see the difference is that the JanusVR Native client still shows references to the default scene carpet, which I cannot edit/remove in Vesta, nor do I see how to edit the js code that seems to reinject that extra JML when opened in Native.

Basically, how can I get the JML to match between Vesta and Native?

**Comment by JamesMcCrae** [comment]
That's odd, I get the same result with either - blank/empty rooms viewing either through Vesta/JanusWeb or Janus native.  I also don't get a crash in either case.

Looking at your room code, you have this script https://vesta.janusvr.com/files/admin/default_webspace/script.js :

```
room.onLoad = function() 
{
	var timetest = room.url.length % 10;
	room.loadNewAsset("Image", {id:"carpet", src:"https://vesta.janusvr.com/files/admin/default_webspace/carpets/carpet" + timetest + ".jpg"})
	room.createObject("Object", {id:"carpet", js_id:"carpet", image_id:"carpet"});
	
	if (room.objects['Scan_id_forJS'])
	{
		room.objects["Scan_id_forJS"].col = Vector(0.5+0.5*Math.random(), 0.5+0.5*Math.random(), 0.5+0.5*Math.random());
	}
}
```

you should replace that `createObject` call to use a plane for id instead:

`room.createObject("Object", {id:"plane", js_id:"carpet", image_id:"carpet"});`

**Comment by cfravel** [comment]
Perhaps I don't understand the way to create a scene in Vesta and then work with it alternately in Vesta and Native.   Or how to create an essentially empty scene to work with, just a ground plane grid and skybox.   I don't see how to edit and persist changes to the javascript, like the one you suggested above, into my Vesta scene.  I don't see how to edit the js in Vesta, nor to save it to the Vesta scene from native.
Am i supposed to manually make a copy of the script that i see in Native, edit that copy, save it in my_files, and change the AssetScript line to point at it?

Here is what I did:

1. I cloned the following scene:
https://vesta.janusvr.com/carlfravel/skybox-and-ground-grid-template

2. I removed the comments from above Firefox, leaving only the one AssetScript, the Room with visible=true, and one object, the vesta_thumbnail_camera.
------------------
<FireBoxRoom>
	<Assets>
		
        	
		<AssetScript src="https://vesta.janusvr.com/files/admin/default_webspace/script.js" />
	</Assets>
	<Room use_local_asset="room_plane" visible="true" pos="-5 0 -3" ambient="1 1 1" pbr="true">
		
      <!	-- this will mark where your thumbnail will be taken and from what angle. If you leave this line out, it will be 1.6 meters above your room position -->
		<Object js_id="vesta_thumbnail_camera" position="-10 2 -3" fwd="0.542248 0 -0.840219" />
	</Room>
</FireBoxRoom>
-----------------

3. I saved the new clone in Vesta.

4. I opened the same URL in Native

5. I see different JML in Native.   Janus native is adding items into my JML.
-----------------
<FireBoxRoom>
<Assets>
<AssetImage id="carpet" src="https://vesta.janusvr.com/files/admin/default_webspace/carpets/carpet9.jpg" />
<AssetObject id="carpet" src="carpet" />
<AssetScript src="https://vesta.janusvr.com/static/js/live_update.js" />
<AssetScript src="https://vesta.janusvr.com/files/admin/scripts/vesta_comment_inject.js" />
<AssetScript src="https://vesta.janusvr.com/files/admin/default_webspace/script.js" />
</Assets>
<Room use_local_asset="room_plane" pos="-5 0 -3">
<Object id="carpet" js_id="carpet" image_id="carpet" />
<Object js_id="vesta_thumbnail_camera" xdir="-0.840219 0 -0.542248" zdir="0.542248 0 -0.840219" />
</Room>
</FireBoxRoom>
-----------------

6. Is there documentation on how to do what I'm trying to do:  Create a simple template file stored in Vesta that, when opened in Native, has the same JML?

7. Is there documentation about how to use my_files to do things like store a simplified set of js that will be in effect for Native?

8. Is it possible to have the same JML in both Vesta and Native?

**Comment by JamesMcCrae** [comment]
Regarding your first question - on changes to scripts that persist.  You can edit script files in Vesta using its web interface, see the following (Scripts->(edit)->Save Script):

![image](https://user-images.githubusercontent.com/1403837/53747367-18cfaa00-3e71-11e9-8943-2cc4a4633595.png)

You can also directly edit scripts in Janus native, see the following (Ctrl+U or Menu->Window->Code Editor->(edit)->Save):

![image](https://user-images.githubusercontent.com/1403837/53747429-3e5cb380-3e71-11e9-9539-95700624126e.png)

**Comment by JamesMcCrae** [comment]
Files authored on Vesta should have the same JML as Janus native client.  However, Vesta "auto-injects" some specific scripts to add functionality to Vesta-hosted pages (this functionality is for spatialized comments, and counting number of users on public pages so people can find each other).  These scripts will appear in the Janus native client, but Vesta's editing interface will not show them:

```
<AssetScript src="https://vesta.janusvr.com/static/js/live_update.js"/>
<AssetScript src="https://vesta.janusvr.com/files/admin/scripts/vesta_comment_inject.js"/>
```

**Comment by JamesMcCrae** [comment]
Regarding: "Is there documentation about how to use my_files to do things like store a simplified set of js that will be in effect for Native?"

If you are authoring with Janus native, it is simple as creating a new script in your local workspace (tab icon with "+" in the Code Editor):

![image](https://user-images.githubusercontent.com/1403837/53748215-e626b100-3e72-11e9-8bc8-649f08081c83.png)

Just make sure you are not editing the Vesta-injected scripts mentioned in the previous comment: live_update.js and vesta_comment_inject.js.

If you want to use the same script in all instances of your template, the script file can be hosted anywhere, and you just need to add the AssetScript tag in the markup of your template.

This script will be visible and work in Janus native.

It is important to note that the Janus native client does not allow for direct editing of room code or scripts directly that are located remotely - this would require logging into Vesta through the native client, which is not an implemented feature.  

Generally when editing a room with Janus native, you create a local workspace (a directory on your local drive), and do editing from within there - edits are saved directly to the local filesystem.  Then when done editing, you can upload these files anywhere, e.g. to Vesta.  

What may be causing a bit of confusion is the "persistent edits" premium feature in Vesta, which monitors server activity for a specific room and allows "interactive editing" (by this I mean clients who make room edits like adding objects and moving them about) to save to the markup - this is a specific feature of Vesta.

**Comment by Spyduck** [comment]
I'll add some (slightly verbose) clarification on what Vesta is doing under the hood.

When you save JML using Vesta's edit page, your JML should remain intact on the edit page, but you still get a slightly different version when viewing the page:

- Vesta parses through the source to find the FireBoxRoom tags, discarding the rest
- Vesta searches for the following AssetScripts, then deletes them if they're fonud
- Vesta then (re-)adds these two AssetScripts:
1. https://vesta.janusvr.com/static/js/live_update.js OR https://vesta.janusvr.com/static/js/live_update_no_party.js (depending on the private setting) - this helps us to differentiate between people actually exploring the pages in one of our clients, vs bots scraping the site
2. https://vesta.janusvr.com/files/admin/scripts/vesta_comment_inject.js - this is for the positional/"floating comments" that are made from Vesta's view page while exploring in the client.
*(note: The second script can be disabled by clicking "Prevent Floating Comments", and in retrospect, not every page needs it to be injected, so this should only happen when a floating comment has been made. I'll make this change later)*
- When you view the page, this modified version is embedded rather than the original source, and this is what you end up seeing in native.

Persistence ("Sandbox" in the edit page UI) is another thing entirely, where essentially the same thing happens but each edit you make (in native or web) that syncs over the network gets edited into the parsed JML. **Sandbox mode will overwrite your JML as edits are made.** You can create a backup of a space with Sandbox enabled if you want to preserve it ("Save and Backup Webspace")

If you want a persistent room to save changes made in JavaScript you need to make sure an object has a js_id and is synced:

```javascript
var cube = room.createObject("Object", {id:"cube", js_id:"cube_js_id_here"});
cube.sync = true; // setting this to true marks it to be synced over the network and resets to false by the next frame
```

**Comment by cfravel** [comment]
Wow, thanks for the great information, James!
